xcp_d/utils/utils.py

```diff
@@ -463,6 +463,36 @@ def denoise_with_nilearn(
         sample_mask=sample_mask,
         t_r=TR,
     )
+    # Replace any high-motion volumes at the beginning or end of the run with the closest
+    # low-motion volume's data.
+    # From https://stackoverflow.com/a/48106843/2589328
+    outlier_idx = sorted(set(np.where(~sample_mask)[0]))
+    if outlier_idx:
+        gaps = [[s, e] for s, e in zip(outlier_idx, outlier_idx[1:]) if s + 1 < e]
+        edges = iter(outlier_idx[:1] + sum(gaps, []) + outlier_idx[-1:])
+        consecutive_outliers_idx = list(zip(edges, edges))
```

Comment:
This is hard for me to follow intuitively, but I tested it a bunch and it seems to work fine.